src/include/processor/operator/recursive_extend/path_scanner.h

```diff
@@ -8,85 +8,116 @@ namespace kuzu {
- * destination nodes (or the BFSMorsel::isAllDstTarget() function which indicates that every node is
- * a possible destination).
+ * FrontierScanner scans all dst nodes from k'th frontier. To identify the
+ * destination nodes in the k'th frontier, we use a semi mask that marks the destination nodes (or
+ * targetDstOffsets.empty() which indicates that every node is a possible destination).
  */
-struct FixedLengthPathScanner {
-    using nbrs_t = std::vector<common::offset_t>*;
-    // FixedLengthPathScanner scans from bfsMorsel frontiers and semi mask.
-    const BaseBFSMorsel& bfsMorsel;
+struct FrontierScanner {
+    std::vector<Frontier*> frontiers;
+    const std::unordered_set<common::offset_t>& targetDstOffsets;
     // Number of extension performed during recursive join
-    size_t numRels;
-    std::vector<common::offset_t> path;
-
-    // DFS states
-    std::stack<nbrs_t> nbrsStack;
-    std::stack<int64_t> cursorStack;
+    size_t numExtensions;
 
     size_t lastFrontierCursor;
+    common::offset_t currentDstOffset;
+
+    FrontierScanner(
+        const std::unordered_set<common::offset_t>& targetDstOffsets, size_t numExtensions)
+        : targetDstOffsets{targetDstOffsets}, numExtensions{numExtensions}, lastFrontierCursor{0},
+          currentDstOffset{common::INVALID_OFFSET} {}
 
-    FixedLengthPathScanner(const BaseBFSMorsel& bfsMorsel, size_t numRels)
-        : bfsMorsel{bfsMorsel}, numRels{numRels} {
-        path.resize(numRels + 1);
-        resetState();
+    size_t scan(common::table_id_t tableID, common::ValueVector* pathVector,
+        common::ValueVector* dstNodeIDVector, common::sel_t& offsetVectorPos,
+        common::sel_t& dataVectorPos);
+
+    void resetState(const BaseBFSMorsel& bfsMorsel);
+
+private:
+    virtual void initScanFromDstOffset() {}
+    virtual void scanFromDstOffset(common::table_id_t tableID, common::ValueVector* pathVector,
+        common::ValueVector* dstNodeIDVector, common::sel_t& offsetVectorPos,
+        common::sel_t& dataVectorPos) {
+        assert(offsetVectorPos < common::DEFAULT_VECTOR_CAPACITY);
+        dstNodeIDVector->setValue<common::nodeID_t>(
+            offsetVectorPos++, common::nodeID_t{currentDstOffset, tableID});
     }
+};
 
-    size_t scanPaths(common::table_id_t tableID, common::ValueVector* pathOffsetVector,
-        common::ValueVector* pathDataVector, common::ValueVector* dstNodeIDVector,
-        common::sel_t& offsetVectorPos, common::sel_t& dataVectorPos);
+/*
+ * FrontierWithEdgeScanner scans all paths of a fixed length k. This is done by starting
+ * a backward traversals from only the destination nodes in the k'th frontier (assuming the first
+ * frontier has index 0) over the backwards edges stored between the frontiers stored in the
+ * BFSMorsel that was used to store the data related to the BFS that was computed in the
+ * RecursiveJoin operator.
+ */
+struct FrontierWithEdgeScanner : public FrontierScanner {
```

Comment:
My main comment is that I see some inconsistencies in the coding style when building different versions of different classes you have: FrontierScanners, Morsels, Frontiers. Sometimes you name the base class as "BaseX" (e.g., in Morsel) and sometimes just "X" (e.g., Frontier). Sometimes you store some data in a common field in the base class and then reinterpret it (e.g., FrontierScanners reinterpret the frontiers stored in the base class to FrontierWithMultiplicity or FrontierWithEdge) and sometimes you store these fields in the extending classes. 

Also at least for the Morsels, it seemed like templating might be better (and probably more efficient during query processing). I would clean this up to ensure you have the same style in every class. It would things more parallel and easier to read. I didn't think hard about what to do here, but I'm happy to think this through quickly with you.